### PR TITLE
[RM-5553]Create waivers with wildcards

### PR DESCRIPTION
**Pull request by ricleal-fugue**
Added support for wildcards

### CODE REVIEW

File: cmd/createRuleWaiver.go [hunk 1]
```diff
@@ -85,15 +110,30 @@ func NewCreateRuleWaiverCommand() *cobra.Command {
...
+			if !opts.WildcardMode && (containsWildcards(opts.ResourceID) ||
+				containsWildcards(opts.ResourceType) ||
+				containsWildcards(opts.ResourceProvider) ||
+				containsWildcards(opts.ResourceTag)) {
+				return errors.New("'wildcard-mode=false' must only be used for exact matches. No wildcards are allowed")
+			}
+			if !isValidTag(opts.ResourceTag) {
+				return errors.New("'resource-tag' must be of the form 'key:value'")
+			}
+			return nil
```
Comment (curtis-fugue) [inline cmd/createRuleWaiver.go:124]: Just curious - would relying on server-side validation be an option here? Having the client simply pass through the values and asking the server to validate could make some sense.

File: cmd/getRuleWaiver.go [hunk 2]
```diff
@@ -54,13 +54,27 @@ func NewGetRuleWaiverCommand() *cobra.Command {
...
+			} else if *waiver.WildcardMode {
+				item = Item{"RESOURCE_TAG", "*"}
+			} else {
+				// this should never happen
+				item = Item{"RESOURCE_TAG", ""}
+			}
+			// Insert resource-tag after resource-provider
+			index := 11
+			items = append(items[:index+1], items[index:]...)
+			items[index] = item
```
Comment (curtis-fugue) [inline cmd/getRuleWaiver.go:77]: If you created the RESOURCE_TAG Item before line 45 you could avoid the slice manipulation here right?

File: cmd/createRuleWaiver.go [hunk 3]
```diff
@@ -85,15 +110,30 @@ func NewCreateRuleWaiverCommand() *cobra.Command {
...
 	cmd.Flags().StringVar(&opts.Name, "name", "", "Waiver name")
 	cmd.Flags().StringVar(&opts.Comment, "comment", "", "Comment describing the rule waiver purpose")
 	cmd.Flags().StringVar(&opts.RuleID, "rule-id", "", "Rule ID (e.g. FG_R00217, <UUID Custom Rule ID>)")
 	cmd.Flags().StringVar(&opts.EnvironmentID, "environment-id", "", "Environment ID")
-	cmd.Flags().StringVar(&opts.ResourceID, "resource-id", "*", "Resource ID")
+	cmd.Flags().StringVar(&opts.ResourceID, "resource-id", "*", "Resource ID (e.g. resource-123, 'resource-*')")
 	cmd.Flags().StringVar(&opts.ResourceType, "resource-type", "*", "Resource Type (e.g. AWS.S3.Bucket, '*')")
 	cmd.Flags().StringVar(&opts.ResourceProvider, "resource-provider", "*", "Resource Provider (e.g. aws.us-east-1, azure, '*')")
+	cmd.Flags().StringVar(&opts.ResourceTag, "resource-tag", "*", "Resource Tag (e.g. 'env:prod', 'fugue:*', '*')")
+	cmd.Flags().BoolVar(&opts.WildcardMode, "wildcard-mode", true, "Wildcard Mode defaults to True. When doing exact matches with 'resource-id', 'resource-type', 'resource-provider' and 'resource-tag' use 'wildcard-mode=false'")
```
Comment (curtis-fugue) [inline cmd/createRuleWaiver.go:136]:
```suggestion
	cmd.Flags().BoolVar(&opts.WildcardMode, "wildcard-mode", true, "Controls whether glob-style wildcard characters are expanded")
```

I think we can keep the help text here simple and elaborate further in documentation.

File: cmd/createRuleWaiver.go [hunk 4]
```diff
@@ -85,15 +110,30 @@ func NewCreateRuleWaiverCommand() *cobra.Command {
...
 
 	cmd.Flags().StringVar(&opts.Name, "name", "", "Waiver name")
 	cmd.Flags().StringVar(&opts.Comment, "comment", "", "Comment describing the rule waiver purpose")
 	cmd.Flags().StringVar(&opts.RuleID, "rule-id", "", "Rule ID (e.g. FG_R00217, <UUID Custom Rule ID>)")
 	cmd.Flags().StringVar(&opts.EnvironmentID, "environment-id", "", "Environment ID")
-	cmd.Flags().StringVar(&opts.ResourceID, "resource-id", "*", "Resource ID")
+	cmd.Flags().StringVar(&opts.ResourceID, "resource-id", "*", "Resource ID (e.g. resource-123, 'resource-*')")
 	cmd.Flags().StringVar(&opts.ResourceType, "resource-type", "*", "Resource Type (e.g. AWS.S3.Bucket, '*')")
 	cmd.Flags().StringVar(&opts.ResourceProvider, "resource-provider", "*", "Resource Provider (e.g. aws.us-east-1, azure, '*')")
+	cmd.Flags().StringVar(&opts.ResourceTag, "resource-tag", "*", "Resource Tag (e.g. 'env:prod', 'fugue:*', '*')")
```
Comment (curtis-fugue) [inline cmd/createRuleWaiver.go:135]:
```suggestion
	cmd.Flags().StringVar(&opts.ResourceTag, "resource-tag", "*", "Resource tag (e.g. 'env:prod', 'env:*')")
```

Just thinking we can keep the examples simple.

File: cmd/createRuleWaiver.go (hunk 1 above)
Comment (ricleal-fugue) [reply cmd/createRuleWaiver.go:124]: What do you suggest? Moving this sort of validation to the API?
Right now, the validation is the API is pretty minimal and this sort of validation is not implemented.

Comment (curtis-fugue) [reply cmd/createRuleWaiver.go:124]: It does seem like the kind of check that should happen in the API, yeah. What do you think @mattm-fugue?

Comment (mattm-fugue) [reply cmd/createRuleWaiver.go:124]: I'm in favor of removing the top check all together. API consumers may want to create exact match waivers with those characters, if we do eventually see them in resource IDs, so we should not prevent them from doing that.

The bottom check is better served in the API, but again, I wonder if it needs to exist. Customers could waive `key-prefix-*` and that would still work for tags like `key-prefix-a: foo` and `key-prefix-b: bar`

**Review by curtis-fugue — APPROVED** [review]

**Pull request merged**